Fixed GE Inverse for noncommutative elements, added test

CerebralPaulZ opened the pull request:
GE multiplication of rows by noncommutative scalars must be applied on left side and immediately, not "last".

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixed inverse_GE for noncommutative elements as inverse_LU was fixed for noncommutative elements in pull https://github.com/sympy/sympy/pull/27437. 
Mentioned as possibility in context of 
https://github.com/sympy/sympy/pull/27422#issuecomment-2571438633
#### Brief description of what is fixed or changed
GE multiplication of rows by noncommutative scalars must be applied on left side and immediately, not "last".

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* matrices
  * Fixed inverse_GE for noncommutative elements.
<!-- END RELEASE NOTES -->

CerebralPaulZ committed:
Fixed GE Inverse for noncommutative elements, added test

GE multiplication of rows by noncommutative scalars must be
applied on left side and immediately, not "last".

CerebralPaulZ committed:
Fixed matrices/test/test_reductions by dropping import

Code Quality failed due to `sympy.simplify.simplify.expand`
imported but unused.

CerebralPaulZ committed:
author: add Paul Scofield to .mailmap

oscarbenjamin commented on .mailmap:
```diff
@@ -1182,6 +1182,7 @@ Pastafarianist <mr.pastafarianist@gmail.com>
 Patrick Lacasse <patrick.m.lacasse@gmail.com>
 Patrick Poitras <acebulf@gmail.com>
 Paul Mandel <paulmandel@google.com>
+Paul Scofield <cerebral.paul@gmail.com>
```
The [commit](https://github.com/sympy/sympy/commit/e4164b0b88610efb1df98791afbd0450177e6ec5.patch) has:
```
Paul Scofield <Cerebral.Paul@gmail.com>
```
You might need:
```
Paul Scofield <cerebral.paul@gmail.com> Paul Scofield <Cerebral.Paul@gmail.com>

```

CerebralPaulZ committed:
author: add Paul Scofield case sensitive to .mailmap

oscarbenjamin commented on .mailmap:
```diff
@@ -1182,6 +1182,8 @@ Pastafarianist <mr.pastafarianist@gmail.com>
 Patrick Lacasse <patrick.m.lacasse@gmail.com>
 Patrick Poitras <acebulf@gmail.com>
 Paul Mandel <paulmandel@google.com>
+Paul Scofield <cerebral.paul@gmail.com>
+Paul Scofield <Cerebral.Paul@gmail.com>
```
There should not be two lines in .mailmap:
https://github.com/sympy/sympy/blob/7450b1a76230162fb2b05bfcde56c2be6095dfae/.mailmap#L129-L148

CerebralPaulZ committed:
author: add Paul Scofield to .mailmap per OJB

oscarbenjamin commented on .mailmap:
```diff
@@ -1182,6 +1182,7 @@ Pastafarianist <mr.pastafarianist@gmail.com>
 Patrick Lacasse <patrick.m.lacasse@gmail.com>
 Patrick Poitras <acebulf@gmail.com>
 Paul Mandel <paulmandel@google.com>
+Paul Scofield <cerebral.paul@gmail.com> Paul Scofield <Cerebral.Paul@gmail.com>
```
I think a second line is needed with
```
Paul Scofield <cerebral.paul@gmail.com> CerebralPaulZ <Cerebral.Paul@gmail.com>
```
I'm not sure where the other name comes from. Might be from GitHub config.

CerebralPaulZ committed:
author: add Paul Scofield to .mailmap 5 CerebralPaulZ

CerebralPaulZ committed:
author: add Paul Scofield to .mailmap minimal 2 CerebralPaulZ

CerebralPaulZ committed:
author: add Paul Scofield to .mailmap reading directions per config

oscarbenjamin commented on sympy/matrices/tests/test_reductions.py:
```diff
@@ -269,6 +269,43 @@ def test_rref():
+    #This result also agrees with Block inverse, upon expansion:
+    assert m_inverse_calculated == m.inverse_BLOCK().expand()
+
+    # More definitively, we expect the products of m and its inverse (in either order)
+    # to be the identity matrix. However, the results are not simple and
+    # because sympy simplification is limited for noncommutative expressions,
+    # current simplify() can only usefully apply expand() in this context.
+    # Consequently (in each case) only 2 of the 4 elements of the
+    # resulting product matrix can be asserted to be 1 or 0.
+    m_m_inverse = (m*m_inverse_calculated).expand()
```
We can get close but it just can't quite do it:
```
In [25]: e = I[1,0]

In [29]: factor_nc(factor_nc(e.expand() - c*a**-1) + c*a**-1)
Out[29]: 
⎛                                 -1⎞      
⎜    ⎛   -1      ⎞ ⎛     -1      ⎞  ⎟    -1
⎝1 + ⎝c⋅a  ⋅b - d⎠⋅⎝- c⋅a  ⋅b + d⎠  ⎠⋅c⋅a
```
I'm not sure how to extract the minus sign.

I think `factor_nc` should extract the minus sign here:
```python
In [31]: factor_nc(e.expand() - c*a**-1)
Out[31]: 
                             -1      
⎛   -1      ⎞ ⎛     -1      ⎞      -1
⎝c⋅a  ⋅b - d⎠⋅⎝- c⋅a  ⋅b + d⎠  ⋅c⋅a  
```

CerebralPaulZ replied:
Resolved in the latest commit b7bf5a8: Applied factor_nc(), raising successful verifications from 2 to 3 of 4
elements each. Tightened comments.

CerebralPaulZ committed:
matrices: Added factor_nc() in test_inverse_GE_noncommutative

Applied factor_nc(), raising successful verifications from 2 to 3 of 4
elements each. Introduced "Schur Complement" variable S, better to match
Wikipedia article. Capitalized all noncommutative symbols. Tightened comments.
Standardized spacing, parentheses, etc.

CerebralPaulZ commented on sympy/matrices/tests/test_reductions.py:
```diff
@@ -269,6 +269,43 @@ def test_rref():
+    m_inverse_expected = Matrix([
+                    [a**(-1) + a**(-1)*b*(-c*a**(-1)*b + d)**(-1)*c*a**(-1), -a**(-1)*b*(-c*a**(-1)*b + d)**(-1)],
+                    [-(-c*a**(-1)*b + d)**(-1)*c*a**(-1), (-c*a**(-1)*b + d)**(-1)]])
+    assert m_inverse_calculated == m_inverse_expected
```
Resolved in the latest commit b7bf5a8 today: Applied factor_nc(), raising successful verifications from 2 to 3 of 4
elements each. Standardized spacing, parentheses, etc.

Even capitalized all noncommutative symbols (to mirror Wikipedia even more precisely).

CerebralPaulZ reviewed (commented):
All issues raised in comments should be resolved in the latest commit b7bf5a8 today, as highlighted in latest comment of each thread.

Capitalizing noncommutative symbols and lowercasing commutative symbols seems a helpful soft standard, going forward.

oscarbenjamin commented:
Looks good. Thanks!
